src/main/java/net/minecraftforge/client/ForgeHooksClient.java

```diff
@@ -1189,106 +1183,25 @@ public static ResourceLocation getShaderImportLocation(String basePath, boolean
         return new ResourceLocation(loc.getNamespace(), normalised);
     }
 
-    public static CreativeModeTab.DisplayItemsGenerator onCreativeModeTabBuildContents(CreativeModeTab tab, CreativeModeTab.DisplayItemsGenerator originalGenerator) {
-        Set<CreativeModeTabEvent.DisplayItemsAdapter> customGenerators = new HashSet<>();
-        CreativeModeTabEvent.BuildContents event = ModLoader.get().postEventWithReturn(new net.minecraftforge.event.CreativeModeTabEvent.BuildContents(tab, customGenerators));
-
-        record ModificationEntry(CreativeModeTab.TabVisibility tabVisibility, ItemStack before, ItemStack after) {}
-
-        return (featureFlagSet, output, hasPermissions) -> {
-            final Map<ItemStack, CreativeModeTab.TabVisibility> originalEntries = ItemStackMap.createTypeAndTagLinkedMap();
-            final Map<ItemStack, ModificationEntry> entries = ItemStackMap.createTypeAndTagMap();
-
-            originalGenerator.accept(featureFlagSet, originalEntries::put, hasPermissions);
-            customGenerators.forEach(generator -> generator.accept(featureFlagSet, (stack, visibility, before, after) -> {
-                if (stack.getCount() != 1)
-                    throw new IllegalArgumentException("The stack count must be 1");
-
-                entries.put(stack, new ModificationEntry(visibility, before, after));
-            }, hasPermissions));
-
-            // We switch to the stacks tag here, as we need to be able to compare and hash them for the graph!
-            // We could not handle this with an open hashset since the builder does not support this internally.
-            final MutableGraph<CompoundTag> graph = GraphBuilder.directed().nodeOrder(ElementOrder.insertion()).build();
-            final Map<CompoundTag, ItemStack> tagToStack = Maps.newHashMap();
-
-            CompoundTag lastStackTag = null;
-            for (ItemStack stack : originalEntries.keySet())
-            {
-                final CompoundTag tag = stack.serializeNBT();
-                tagToStack.put(tag, stack);
-                if (!stack.isEmpty())
-                {
-                    graph.addNode(tag);
-                }
-                if (lastStackTag != null)
-                {
-                    graph.putEdge(lastStackTag, tag);
-                }
-                lastStackTag = tag;
+    public static void onCreativeModeTabBuildContents(CreativeModeTab tab, CreativeModeTab.DisplayItemsGenerator originalGenerator, FeatureFlagSet flags, CreativeModeTab.Output output, boolean hasPermisions) {
+        final var entries = new MutableHashedLinkedList<ItemStack, CreativeModeTab.TabVisibility>(ItemStackLinkedSet.TYPE_AND_TAG,
+            (key, left, right) -> {
+                //throw new IllegalStateException("Accidentally adding the same item stack twice " + key.getDisplayName().getString() + " to a Creative Mode Tab: " + tab.getDisplayName().getString());
```

Comment:
Remove the commented-out code?

Reply:
I kept it in as more of a comment about what vanilla does.